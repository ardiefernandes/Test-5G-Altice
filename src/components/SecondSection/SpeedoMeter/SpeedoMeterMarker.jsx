import PropTypes from "prop-types";
import React, { useCallback, useEffect, useRef } from "react";

const SpeedoMeterMarker = ({ radius = 45, lineWidth = 4, value }) => {
  const canvasRef = useRef();
  const circle = useRef({
    canvas: null,
    context: null,
    x: 0,
    y: 0,
    radius,
    start: Math.PI / 1.5,
    end: Math.PI * 1.65,
  });

  useEffect(() => {
    try {
      if (!circle.current.canvas) {
        circle.current.canvas = canvasRef.current;
        circle.current.context = circle.current.canvas.getContext("2d");

        const gradient = circle.current.context.createLinearGradient(
          0,
          radius * 2,
          0,
          0
        );

        gradient.addColorStop(0, "red");
        gradient.addColorStop(0.5, "yellow");
        gradient.addColorStop(1, "#3282F9");
        circle.current.x = circle.current.canvas.width / 2;
        circle.current.y = circle.current.canvas.height / 2;
        circle.current.context.lineWidth = lineWidth;
        circle.current.context.strokeStyle = gradient;
        circle.current.context.shadowOffsetX = 0;
        circle.current.context.shadowOffsetY = 0;
        circle.current.context.shadowBlur = lineWidth * 10;
        circle.current.context.shadowColor = gradient;
      }

      drawCircle(value);
    } catch (e) {
      // Jest doesn't supports canvas snapshots, so exceptions
      // need to be caught when it makes a snapshot and tries to
      // perform the code above to draw the canvas circle
    }
  }, [value]);

  const drawCircle = useCallback((value) => {
    const end = value
      ? (value * circle.current.end) / 100
      : circle.current.start;

    circle.current.context.clearRect(
      0,
      0,
      circle.current.canvas.width,
      circle.current.canvas.height
    );
    circle.current.context.beginPath();
    circle.current.context.arc(
      circle.current.x,
      circle.current.y,
      circle.current.radius,
      circle.current.start,
      end > circle.current.start ? end : circle.current.start,
      false
    );

    circle.current.context.stroke();
  });

  return (
    <canvas
      ref={canvasRef}
      width={radius * 2 + lineWidth + 6}
      height={radius * 2 + lineWidth + 6}
    ></canvas>
  );
};

SpeedoMeterMarker.propTypes = {
  radius: PropTypes.number,
  lineWidth: PropTypes.number,
  value: PropTypes.number,
};

export default SpeedoMeterMarker;
