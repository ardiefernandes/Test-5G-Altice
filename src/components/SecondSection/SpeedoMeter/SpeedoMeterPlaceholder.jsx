import PropTypes from "prop-types";
import React, { memo, useEffect, useRef } from "react";

const SpeedoMeterPlaceholder = ({
  radius = 45,
  lineWidth = 4,
  start = Math.PI / 1.5,
  end = Math.PI * 1.65,
}) => {
  const canvasRef = useRef();

  useEffect(() => {
    try {
      const canvas = canvasRef.current;
      const context = canvas.getContext("2d");

      context.lineWidth = lineWidth;
      context.strokeStyle = "#A8A8A8";
      context.shadowOffsetX = 0;
      context.shadowOffsetY = 0;
      context.shadowBlur = 3;
      context.shadowColor = "#A8A8A8";

      context.beginPath();
      context.arc(
        canvas.width / 2,
        canvas.height / 2,
        radius,
        start,
        end,
        false
      );
      context.stroke();
    } catch (e) {
      // Jest doesn't supports canvas snapshots, so exceptions
      // need to be caught when it makes a snapshot and tries to
      // perform the code above to draw the canvas circle
    }
  }, []);

  return (
    <canvas
      ref={canvasRef}
      width={radius * 2 + lineWidth + 6}
      height={radius * 2 + lineWidth + 6}
    ></canvas>
  );
};

SpeedoMeterPlaceholder.propTypes = {
  radius: PropTypes.number,
  lineWidth: PropTypes.number,
  start: PropTypes.number,
  end: PropTypes.number,
};

export default memo(SpeedoMeterPlaceholder);
