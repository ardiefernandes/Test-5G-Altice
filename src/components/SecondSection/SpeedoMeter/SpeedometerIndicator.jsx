import PropTypes from "prop-types";
import React from "react";

const SpeedometerIndicator = ({ label, speed, measureScale }) => {
  return (
    <div className="speedometer-text">
      {label ? <label className="speedometer-label">{label}</label> : null}
      <span className="speedometer-indicator">
        {Math.round(speed * 100) / 100 +
          (measureScale ? " " + measureScale : "")}
      </span>
    </div>
  );
};

SpeedometerIndicator.propTypes = {
  label: PropTypes.string,
  speed: PropTypes.number,
  measureScale: PropTypes.string,
};

export default SpeedometerIndicator;
