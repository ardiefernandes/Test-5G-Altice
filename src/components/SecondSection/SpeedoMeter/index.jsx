import PropTypes from "prop-types";
import React, { useCallback, useEffect, useRef, useState } from "react";
import SpeedometerIndicator from "./SpeedometerIndicator";
import SpeedoMeterMarker from "./SpeedoMeterMarker";
import SpeedoMeterPlaceholder from "./SpeedoMeterPlaceholder";

const SpeedoMeter = ({
  label = "",
  size = 120,
  lineWidth = 4,
  minSpeed = 0,
  maxSpeed = 100,
  measureScale = "",
  range = [0, 100],
  // flag added to avoid non-stop recursivity during the tests
  test = false,
}) => {
  const timeoutRef = useRef();
  const [speed, setSpeed] = useState(0);

  useEffect(() => {
    if (!test) {
      randomSpeed();
    }

    return () => {
      clearTimeout(timeoutRef.current);
    };
  }, []);

  const randomSpeed = useCallback(() => {
    const difference = maxSpeed - minSpeed;
    const rand = Math.random();

    setSpeed(minSpeed + rand * difference);

    if (!test) {
      timeoutRef.current = setTimeout(() => randomSpeed(), 500);
    }
  });

  return (
    <div
      className="speedometer"
      style={{ height: size + 10, width: size + 10 }}
    >
      <SpeedometerIndicator
        label={label}
        speed={speed}
        measureScale={measureScale}
      />
      <SpeedoMeterPlaceholder radius={size / 2} lineWidth={lineWidth} />
      <SpeedoMeterMarker
        value={(speed / (range[1] - range[0])) * 100}
        radius={size / 2}
        lineWidth={lineWidth}
      />
    </div>
  );
};

SpeedoMeter.propTypes = {
  label: PropTypes.string,
  size: PropTypes.number,
  lineWidth: PropTypes.number,
  minSpeed: PropTypes.number,
  maxSpeed: PropTypes.number,
  measureScale: PropTypes.string,
  range: PropTypes.array,
  test: PropTypes.bool,
};

export default SpeedoMeter;
