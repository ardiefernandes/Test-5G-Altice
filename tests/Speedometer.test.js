import React from "react";
import renderer from "react-test-renderer";
import SpeedoMeter from "../src/components/SecondSection/SpeedoMeter";

it("correctly renders the speedometer for 3G", () => {
  const component = renderer.create(
    <SpeedoMeter
      label="3G"
      size={380}
      lineWidth={12}
      minSpeed={0.1}
      maxSpeed={2}
      measureScale="Mbps"
      range={[0, 1000]}
      test={true}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("correctly renders the speedometer for 4G", () => {
  const component = renderer.create(
    <SpeedoMeter
      label="4G"
      size={420}
      lineWidth={14}
      minSpeed={400}
      maxSpeed={500}
      measureScale="Mbps"
      range={[0, 1000]}
      test={true}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("correctly renders the speedometer for 5G", () => {
  const component = renderer.create(
    <SpeedoMeter
      label="5G"
      size={450}
      lineWidth={15}
      minSpeed={700}
      maxSpeed={1000}
      measureScale="Mbps"
      range={[0, 1000]}
      test={true}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
