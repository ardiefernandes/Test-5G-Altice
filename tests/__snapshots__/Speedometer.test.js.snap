// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly renders the speedometer for 3G 1`] = `
<div
  className="speedometer"
  style={
    {
      "height": 390,
      "width": 390,
    }
  }
>
  <div
    className="speedometer-text"
  >
    <label
      className="speedometer-label"
    >
      3G
    </label>
    <span
      className="speedometer-indicator"
    >
      0 Mbps
    </span>
  </div>
  <canvas
    height={398}
    width={398}
  />
  <canvas
    height={398}
    width={398}
  />
</div>
`;

exports[`correctly renders the speedometer for 4G 1`] = `
<div
  className="speedometer"
  style={
    {
      "height": 430,
      "width": 430,
    }
  }
>
  <div
    className="speedometer-text"
  >
    <label
      className="speedometer-label"
    >
      4G
    </label>
    <span
      className="speedometer-indicator"
    >
      0 Mbps
    </span>
  </div>
  <canvas
    height={440}
    width={440}
  />
  <canvas
    height={440}
    width={440}
  />
</div>
`;

exports[`correctly renders the speedometer for 5G 1`] = `
<div
  className="speedometer"
  style={
    {
      "height": 460,
      "width": 460,
    }
  }
>
  <div
    className="speedometer-text"
  >
    <label
      className="speedometer-label"
    >
      5G
    </label>
    <span
      className="speedometer-indicator"
    >
      0 Mbps
    </span>
  </div>
  <canvas
    height={471}
    width={471}
  />
  <canvas
    height={471}
    width={471}
  />
</div>
`;
